.github/workflows/release.yml

```diff
@@ -0,0 +1,315 @@
+
+    - name: Get project name
+      run: |
+        if [ ! -f "${PROJECT_DIR}/Cargo.toml" ]; then
+          echo "Error: ${PROJECT_DIR}/Cargo.toml does not exist"
+          exit 1
+        fi
+
+        PROJECT_NAME=$(sed -n 's/^name = "\(.*\)"/\1/p' "${PROJECT_DIR}/Cargo.toml" | head -n1)
+        echo "Project Name: $PROJECT_NAME"
+
+        if [[ "${PROJECT_INDEX_URL}" == *"${PROJECT_NAME}" ]]; then
+          echo "PROJECT_NAME=${PROJECT_NAME}" >> $GITHUB_ENV
+        else
+          echo "Project name does not match the index URL [${PROJECT_INDEX_URL}]."
+          exit 1
+        fi
+
+    - name: Version introspection
+      run: |
+        PROJECT_VERSION=$(sed -n 's/^version = "\(.*\)"/\1/p' "${PROJECT_DIR}/Cargo.toml" | head -n1)
+        echo "PROJECT_VERSION=${PROJECT_VERSION}" >> $GITHUB_ENV
+        echo "Project version: ${PROJECT_VERSION}"
+
+        if [[ ! "${PROJECT_VERSION}" =~ ^[0-9]+\.[0-9]+\.[0-9]+(-[a-zA-Z0-9.-]+)?$ ]]; then
+          echo "Project version is not a valid semver: ${PROJECT_VERSION}"
+          exit 1
+        fi
+
+        if [[ "${PROJECT_VERSION}" == *-* ]]; then
+          echo "Project version [${PROJECT_VERSION}] is a pre-release."
+          echo "IS_PRERELEASE=true" >> $GITHUB_ENV
+        else
+          echo "Project version [${PROJECT_VERSION}] is not a pre-release."
+          echo "IS_PRERELEASE=false" >> $GITHUB_ENV
+        fi
+
+    - name: Check if crates.io release exists
+      run: |
+        CRATE_VERSIONS=$(curl -sL "${PROJECT_INDEX_URL}" | jq -r '.vers')
+        echo -e "Already published versions:\n${CRATE_VERSIONS}"
+
+        if echo "${CRATE_VERSIONS}" | grep -q "^${PROJECT_VERSION}$"; then
+          echo "Project version [${PROJECT_VERSION}] has already been released."
+          echo "SHOULD_PUBLISH_CRATE=false" >> $GITHUB_ENV
+        else
+          echo "Project version [${PROJECT_VERSION}] has not been released."
+          echo "SHOULD_PUBLISH_CRATE=true" >> $GITHUB_ENV
+        fi
+
+    - name: Check if GitHub release exists
+      run: |
+        RELEASED_VERSIONS=$(git tag -l "v[0-9]*.[0-9]*.[0-9]*")
+        echo -e "Already released versions:\n${RELEASED_VERSIONS}"
+
+        if echo "${RELEASED_VERSIONS}" | grep -q "^v${PROJECT_VERSION}$"; then
+          echo "Project version [${PROJECT_VERSION}] has already been released."
+          echo "SHOULD_RELEASE_BINARY=false" >> $GITHUB_ENV
+        else
+          echo "Project version [${PROJECT_VERSION}] has not been released."
+          echo "SHOULD_RELEASE_BINARY=true" >> $GITHUB_ENV
+        fi
+
+  crate:
+    runs-on: ubuntu-latest
+    needs: check
+    if: needs.check.outputs.should_publish_crate == 'true'
+    env:
+      PROJECT_NAME: ${{ needs.check.outputs.project_name }}
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v3
+
+    - name: Install rust
+      uses: dtolnay/rust-toolchain@stable
+
+    - uses: Swatinem/rust-cache@v2
```

Comment:
cache seems unnecessary for the publish step

Reply:
Indeed, it was helpful for testing, but unnecessary here!